File: leetcode/medium/109-sorted-ll-bst/rec2.py
"""
  Problem 109 - Sorted LL to BST

  Runtime - 136 ms
  Memory Usage - 17.8 MB
"""
from typing import Optional


# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next


class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
    if not head:
      return None

    if not head.next:
      return TreeNode(head.val)

    p, mid = self.middleNode(head)

    newNode = TreeNode(mid.val)
    if p:
      p.next = None
      newNode.left = self.sortedListToBST(head)

    newNode.right = self.sortedListToBST(mid.next)

    return newNode

  def middleNode(self, head: Optional[ListNode]):
    if not head:
      return None

    slow, fast = head, head
    prev = None

    while fast.next is not None and fast.next.next is not None:
      prev = slow
      slow = slow.next
      fast = fast.next.next

    return (prev, slow)

File: leetcode/medium/109-sorted-ll-bst/rec2.cpp
/**
 * @file rec2.cpp
 * @author Vortexx2
 * @brief Problem 109 - Convert Sorted LL to BST
 * @date 02-10-2021
 * 
 * Runtime - 47 ms
 * Memory Usage - 30.9 MB
 */
#include <algorithm>
#include <iostream>
#include <utility>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode *sortedListToBST(ListNode *head) {
    if (head == nullptr) return nullptr;
    if (head->next == nullptr) return new TreeNode(head->val);

    ListNode *mid, *p;
    tie(p, mid) = middleNode(head);

    TreeNode *newNode = new TreeNode(mid->val);
    if (p) {
      p->next = nullptr;
      newNode->left = sortedListToBST(head);
    }
    newNode->right = sortedListToBST(mid->next);

    return newNode;
  }

  pair<ListNode *, ListNode *> middleNode(ListNode *head) {
    if (!head) return {nullptr, nullptr};
    ListNode *p = head, *q = head;
    ListNode *prev = nullptr;

    while (q->next && q->next->next) {
      prev = p;
      p = p->next;
      q = q->next->next;
    }

    return {prev, p};
  }
};